code/modules/admin/banjob.dm

```diff
@@ -38,89 +118,164 @@ var/jobban_keylist[0]		//to store the keys & ranks
+	// No database. Weee.
+	if (!establish_db_connection(dbcon))
+		error("Database connection failed. Reverting to the legacy ban system.")
+		log_misc("Database connection failed. Reverting to the legacy ban system.")
+		config.ban_legacy_system = 1
+		jobban_loadbanfile()
+		return
 
-			jobban_keylist.Add("[ckey] - [job]")
+	// All jobbans in one query. Because we don't actually care.
+	var/DBQuery/query = dbcon.NewQuery("SELECT id, ckey, job, reason FROM ss13_ban WHERE isnull(unbanned) AND ((bantype = 'JOB_PERMABAN') OR (bantype = 'JOB_TEMPBAN' AND expiration_time > Now()))")
+	query.Execute()
 
-		//Job tempbans
-		var/DBQuery/query1 = dbcon.NewQuery("SELECT ckey, job FROM ss13_ban WHERE bantype = 'JOB_TEMPBAN' AND isnull(unbanned) AND expiration_time > Now()")
-		query1.Execute()
+	while (query.NextRow())
+		var/ckey = ckey(query.item[2])
+		var/job = query.item[3]
+		var/reason = query.item[4]
 
-		while(query1.NextRow())
-			var/ckey = query1.item[1]
-			var/job = query1.item[2]
+		if (!jobban_keylist[ckey])
+			jobban_keylist[ckey] = list()
 
-			jobban_keylist.Add("[ckey] - [job]")
+		if (!jobban_keylist[ckey][job])
+			// Insert it with 0 time because the expiration of a temp jobban for
+			// MySQL is dependent on the database and query itself. So we can just
+			// politely not care.
+			jobban_keylist[ckey][job] = list(reason, 0)
+		else
+			// Woups. What happened here...?
+			log_and_message_admins("JOBBANS: Duplicate jobban entry in MySQL for [ckey]. Ban ID: #[query.item[1]]")
 
+/**
+ * @name	jobban_savebanfile()
+ * @desc	Saves the current bans into the data/job_full.ban file.
+ *
+ * @return	null
+ */
 /proc/jobban_savebanfile()
-	var/savefile/S=new("data/job_full.ban")
-	S["keys[0]"] << jobban_keylist
+	var/savefile/S = new("data/job_full.ban")
+	S["bans"] << jobban_keylist
 
-/proc/jobban_unban(mob/M, rank)
-	jobban_remove("[M.ckey] - [rank]")
-	jobban_savebanfile()
+/**
+ * @name	jobban_unban()
+ * @desc	Removes a jobban entry from the code and calls the config appropriate
+ *			record removal method. In case of legacy bans, jobban file is updated,
+ *			otherwise, DB queries are ran.
+ *
+ * @param	mob M		The mob holding the ckey to be unbanned.
+ * @param	str rank	The job from which we want to unban the given player.
+ * @param	str ckey	A ckey of the player to be unbanned. This works as a backup
+ *						in case a mob is never sent.
+ * @param	datum/admins holder	The holder of the admin doing the unbanning.
+ *						required to access the DB ban procs.
+ *
+ * @return	null
+ */
+/proc/jobban_unban(mob/M, rank, ckey = null, datum/admins/holder)
+	if (M && M.ckey)
+		ckey = M.ckey
+	else if (!ckey)
+		log_debug("JOBBAN: jobban_unban called without a mob and a backup ckey.")
+		return
 
+	// Check for a player record.
+	var/list/entry = jobban_keylist[ckey]
+	if (entry)
+		// Find the specific ban.
+		var/list/ban = entry[rank]
+		if (ban)
+			// Remove said ban. Lists pass by ref, so this is fine.
+			entry -= rank

```

Comment:
Here, check if entry is an empty list. If yes, just remove the entry entirely by running `jobban_keylist -= ckey`